File: LeetCode/valid-anagram.cpp
// https://leetcode.com/problems/valid-anagram/

class Solution
{
public:
	bool isAnagram(string s, string t)
	{
		int str1Size = s.size();
		int str2Size = t.size();

		if (str1Size != str2Size)
		{
			return false;
		}

		unordered_map<char, int> m;
		unordered_map<char, int>::iterator itr;

		for (int i = 0; i < str1Size; i++)
		{
			itr = m.find(s[i]);

			if (itr != m.end())
			{
				itr->second++;
			}
			else
			{
				m.insert({s[i], 1});
			}
		}

		for (int i = 0; i < str1Size; i++)
		{
			itr = m.find(t[i]);

			if (itr != m.end() && (itr->second > 0))
			{
				itr->second--;
			}
			else
			{
				return false;
			}
		}

		return true;
	}
};

File: LeetCode/Java/validanagram/ValidAnagram.java
package validanagram;

public class ValidAnagram {
	public static void main(String[] args) {
		Solution sol = new Solution();
		boolean result = sol.isAnagram("anagram", "nagaram");
		System.out.println(result);
	}
}

// https://leetcode.com/problems/valid-anagram
//IMPORTANT!! Submit Code Region Begin(Do not remove this line)
class Solution {
	public boolean isAnagram(String s, String t) {
		if (s.length() != t.length()) {
			return false;
		}

		// s and t consist of lowercase English letters
		int[] arr = new int[26];

		for (int i = 0; i < s.length(); i++) {
			arr[s.charAt(i) - 'a']++;
			arr[t.charAt(i) - 'a']--;
		}

		for (int i = 0; i < 26; i++) {
			if (arr[i] != 0) {
				return false;
			}
		}

		return true;
	}
}
//IMPORTANT!! Submit Code Region End(Do not remove this line)
